File: _old_csharp/addons/terrabrush/Scripts/Tools/TextureTool.cs
#if TOOLS
using System;
using System.Collections.Generic;
using System.Linq;
using Godot;

namespace TerraBrush;

public class TextureTool : ToolBase {
    private Dictionary<ImageTexture, Image> _splatmapImagesCache = null;

    public TextureTool(TerraBrush terraBrush) : base(terraBrush) {}

    public override void BeginPaint () {
        base.BeginPaint();

        _splatmapImagesCache = new Dictionary<ImageTexture, Image>();
    }

    public override void EndPaint () {
        base.EndPaint();

        _splatmapImagesCache = null;
    }

    public override void Paint(TerrainToolType toolType, Image brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
        if (_terraBrush.TextureSetIndex == null) {
            return;
        }

        var splatmapIndex = Mathf.FloorToInt(_terraBrush.TextureSetIndex.Value / 4);
        var colorChannel = _terraBrush.TextureSetIndex.Value % 4;
        var numberOfSplatmaps = Mathf.CeilToInt(_terraBrush.TextureSets.TextureSets.Count() / 4.0);

        ForEachBrushPixel(brushImage, brushSize, imagePosition, (imageZoneInfo, pixelBrushStrength) => {
            for (int i = 0; i < numberOfSplatmaps; i++) {
                Color splatmapColor = Colors.Transparent;

                if (i != splatmapIndex) {
                    splatmapColor = new Color(0, 0, 0, 0);
                } else if (colorChannel == 0) {
                    splatmapColor = new Color(1, 0, 0, 0);
                } else if (colorChannel == 1) {
                    splatmapColor = new Color(0, 1, 0, 0);
                } else if (colorChannel == 2) {
                    splatmapColor = new Color(0, 0, 1, 0);
                } else if (colorChannel == 3) {
                    splatmapColor = new Color(0, 0, 0, 1);
                }

                var currentSplatmapTexture = imageZoneInfo.Zone.SplatmapsTexture[i];
                // Since we play with several textures here, the toolbase cannot add the texture to the dirty collection
                _terraBrush.TerrainZones.AddDirtyImageTexture(currentSplatmapTexture);
                AddTextureToUndo(currentSplatmapTexture);

                _splatmapImagesCache.TryGetValue(currentSplatmapTexture, out Image currentSplatmapImage);

                if (currentSplatmapImage == null) {
                    currentSplatmapImage = currentSplatmapTexture.GetImage();
                    _splatmapImagesCache.Add(currentSplatmapTexture, currentSplatmapImage);
                }

                var currentPixel = currentSplatmapImage.GetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y);
                var newValue = new Color(
                    Mathf.Lerp(currentPixel.R, splatmapColor.R, pixelBrushStrength * brushStrength),
                    Mathf.Lerp(currentPixel.G, splatmapColor.G, pixelBrushStrength * brushStrength),
                    Mathf.Lerp(currentPixel.B, splatmapColor.B, pixelBrushStrength * brushStrength),
                    Mathf.Lerp(currentPixel.A, splatmapColor.A, pixelBrushStrength * brushStrength)
                );
                currentSplatmapImage.SetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y, newValue);
            }
        });

        _terraBrush.TerrainZones.UpdateSplatmapsTextures();
    }
}
#endif

File: src/editor_tools/texture_tool.cpp
#include "texture_tool.h"

using namespace godot;

void TextureTool::_bind_methods() {}

TextureTool::TextureTool() {}

TextureTool::~TextureTool() {}

void TextureTool::beginPaint() {
    ToolBase::beginPaint();

    _splatmapImagesCache = std::unordered_map<Ref<ImageTexture>, Ref<Image>>();
}

void TextureTool::endPaint() {
    ToolBase::endPaint();

    _splatmapImagesCache.clear();
}

void TextureTool::paint(TerrainToolType toolType, Ref<Image> brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
    if (_selectedTextureIndex < 0) {
        return;
    }

    int splatmapIndex = (int) Math::floor((float) _selectedTextureIndex / 4);
    int colorChannel = _selectedTextureIndex % 4;
    int numberOfSplatmaps = (int) Math::ceil(_terraBrush->get_textureSets()->get_textureSets().size() / 4.0);

    Color transparentColor = Color(1, 1, 1, 0); // Transparent color

    forEachBrushPixel(brushImage, brushSize, imagePosition, ([&](ImageZoneInfo &imageZoneInfo, float pixelBrushStrength) {
        for (int i = 0; i < numberOfSplatmaps; i++) {
            Color splatmapColor = transparentColor;

            if (i != splatmapIndex) {
                splatmapColor = Color(0, 0, 0, 0);
            } else if (colorChannel == 0) {
                splatmapColor = Color(1, 0, 0, 0);
            } else if (colorChannel == 1) {
                splatmapColor = Color(0, 1, 0, 0);
            } else if (colorChannel == 2) {
                splatmapColor = Color(0, 0, 1, 0);
            } else if (colorChannel == 3) {
                splatmapColor = Color(0, 0, 0, 1);
            }

            Ref<ImageTexture> currentSplatmapTexture = imageZoneInfo.zone->get_splatmapsTexture()[i];
            // Since we play with several textures here, the toolbase cannot add the texture to the dirty collection
            _terraBrush->get_terrainZones()->addDirtyImageTexture(currentSplatmapTexture);
            addTextureToUndo(currentSplatmapTexture);

            Ref<Image> currentSplatmapImage = nullptr;
            if (_splatmapImagesCache.count(currentSplatmapTexture) > 0) {
                currentSplatmapImage = _splatmapImagesCache[currentSplatmapTexture];
            }

            if (currentSplatmapImage.is_null()) {
                currentSplatmapImage = currentSplatmapTexture->get_image();
                _splatmapImagesCache[currentSplatmapTexture] = currentSplatmapImage;
            }

            Color currentPixel = currentSplatmapImage->get_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y);
            Color newValue = Color(
                Math::lerp(currentPixel.r, splatmapColor.r, pixelBrushStrength * brushStrength),
                Math::lerp(currentPixel.g, splatmapColor.g, pixelBrushStrength * brushStrength),
                Math::lerp(currentPixel.b, splatmapColor.b, pixelBrushStrength * brushStrength),
                Math::lerp(currentPixel.a, splatmapColor.a, pixelBrushStrength * brushStrength)
            );
            currentSplatmapImage->set_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y, newValue);
        }
    }));

    _terraBrush->get_terrainZones()->updateSplatmapsTextures();
}

void TextureTool::updateSelectedTextureIndex(int value) {
    _selectedTextureIndex = value;
}